### PR TITLE
Replace ModFrmHilDEltComp with HMFSerPuis

### PR DESCRIPTION
**Pull request by abhijit-mudigonda**

### CODE REVIEW

**Comment by edgarcosta** [comment]
I'm confused about the title (I have not looked at the diff), shouldn'tplp we just replace the associative array of coefficients?

**Comment by j-kieffer** [comment]
Just started reading the diff; looks great so far!
I tend to agree with Edgar. I would prefer to keep the name `ModFrmHilDEltComp` instead of `HMFSerPuis`. It fits better with other Magma names in the package, and from a math point of view, the user cares that it's a component of some HMF, not that it's implemented as a Puiseux series. I'll make more detailed suggestions in a moment

**Comment by abhijit-mudigonda** [comment]
Sure, that makes sense to me. I'll change the name of `HMFSerPuisElt` to `ModFrmHilDEltComp` then. 

There are a couple more commits coming in to improve performance (there are regressions relative to the PR before this, but I think they're related to the calling sites rather than this logic itself) but I wanted to get the main design change up for review first.

**Comment by abhijit-mudigonda** [comment]
Can I leave the parent object of `ModFrmHilDEltComp` as `HMFSerPuis` or should that be renamed as well?

File: ModFrmHilD/HMFSerPuis.m [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+  HMF series associated to different components";
+
+  return cHMFSerPuisElt(Space(f) / Space(g), bb, ShadowSeries(f) / ShadowSeries(g) : 
+    coeff_ring := T`CoefficientRing, prec := prec, prune := true);
+end intrinsic;
+
+/////////////////// HMFSerPuisElt arithmetic helpers /////////////////// 
+
+intrinsic ShadowSeries(f::HMFSerPuisElt) -> RngSerPuisElt
+  {TODO}
```
Comment (j-kieffer) [inline ModFrmHilD/HMFSerPuis.m:577]: Did you want to leave this as a todo?

File: ModFrmHilD/HMFSerPuis.m [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+  return f`ShadowSeries;
+end intrinsic;
+
+intrinsic Series(f::HMFSerPuisElt) -> RngSerPuisElt
+  {TODO}
+  return f`Series;
+end intrinsic;
+
+intrinsic Series(f_ser::RngSerPuisElt, R::HMFSerPuis, bb::RngOrdIdl, prec::RngIntElt) -> RngSerPuisElt
+  {TODO}
```
Comment (j-kieffer) [inline ModFrmHilD/HMFSerPuis.m:600]: Other todos here, just checking if you want to leave them

File: ModFrmHilD/HMFSerPuis.m [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+  g_ser := RngSerPuisZero(R);
+  for nu in FunDomainRepsUpToNorm(R`GRng, bb, prec) do
+    a_nu := Coefficient(f_ser, Depth(R), nu); // this will be assigned if f`Series isn't
+    g_ser +:= RngSerPuisMonomial(R, nu, a_nu);
```
Comment (j-kieffer) [inline ModFrmHilD/HMFSerPuis.m:604]: Same here - `Polynomial` would be better than a loop

File: ModFrmHilD/HMFSerPuis.m [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+end intrinsic;
+
+intrinsic MapCoefficients(m::Map, f::HMFSerPuisElt) -> HMFSerPuisElt
+  {return the ModFrmHilDEltComp where the map acts on the coefficients}
+  R := Parent(f);
+  M := Parent(f)`GRng;
+  g_ser := RngSerPuisZero(R);
+  for nu in FunDomainRepsUpToNorm(M, ComponentIdeal(f), Precision(f)) do
+    a_nu := Coefficient(f, nu);
+    g_ser +:= RngSerPuisMonomial(R, nu, m(a_nu));
```
Comment (j-kieffer) [inline ModFrmHilD/HMFSerPuis.m:634]: Same here - `Polynomial` would be better than a loop

File: ModFrmHilD/HMFSerPuis.m [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+
+intrinsic HMFSerPuisIdentity(Mk::ModFrmHilD, bb::RngOrdIdl) -> HMFSerPuisElt
+  {
+    Returns the identity element in R.
+  }
+  R := GetHMFSerPuis(Parent(Mk), Rationals());
+  return cHMFSerPuisElt(Mk, bb, PuiseuxRing(R)!1);
+end intrinsic;
+
+intrinsic HMFIdentity(Mk::ModFrmHilD, bb::RngOrdIdl) -> HMFSerPuisElt
```
Comment (j-kieffer) [inline ModFrmHilD/HMFSerPuis.m:192]: What's the difference between these two versions of `Identity`?

**Review by j-kieffer — COMMENTED** [review]
Generally I suggest we use a multivariate polynomial ring instead of nested Puiseux series rings. I tried to make detailed suggestions for this.
Other changes we could make:
- drop the `HMFSerPuis` structure
- rename `HMFSerPuisEld` to `ModFrmHilDEltComp` as before
- rename file to `Components.m` for instance, or move back to `Elements.m`
In any case, great work!

File: ModFrmHilD/HMFSerPuis.m [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+  GRng, // ModFrmHilDGradedRing - graded ring of HMFs
+  RngSerPuiss; // Assoc: RngIntElt -> RngSerPuis
+               
+declare attributes HMFSerPuisElt:
+  Parent, // HMFSerPuis
+  Series, // RngSerPuisElt[RngSerPuis[...[RngSerPuis]]...]
+  ShadowSeries, // RngSerPuis[RngSerPuis[...[RngSerPuis]]...]
+  Precision, // RngIntElt - the maximum norm of nn for which coefficients are stored
+  Space, // ModFrmHilD - the HMF space that this HMFSerPuisElt is a component in
+  ComponentIdeal; // RngOrdIdl
```
[the remark this answers is not in the document]
Comment (j-kieffer) [reply ModFrmHilD/HMFSerPuis.m:36]: Actually I'm not even sure that's needed if we choose the basis correctly.

File: ModFrmHilD/HMFSerPuis.m [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+
+  f_ser := PuiseuxRing(R)!0;
+  for nu in FunDomainRepsUpToNorm(M, bb, prec) do
+    f_ser +:= RngSerPuisMonomial(R, nu, coeffs_by_nu[nu]);
+  end for;
+
+  return cHMFSerPuisElt(Mk, bb, f_ser : coeff_ring := coeff_ring, prec := prec);
+end intrinsic;
+
+intrinsic RngSerPuisMonomial(R::HMFSerPuis, nu::FldElt, a_nu::FldElt) -> RngSerPuisElt
```
[the remark this answers is not in the document]
Comment (j-kieffer) [reply ModFrmHilD/HMFSerPuis.m:133]: Edit: the first trick doesn't work, so we have to rescale.

**Pull request merged**